src/set.cr

```diff
@@ -136,17 +136,18 @@ struct Set(T)
     @hash.has_key?(object)
   end
 
-  # Removes the *object* from the set and returns `self`.
+  # Removes the *object* and returns it, otherwise returns `nil`.
   #
   # ```
   # s = Set{1, 5}
   # s.includes? 5 # => true
-  # s.delete 5
+  # s.delete 5    # => 5
   # s.includes? 5 # => false
   # ```
-  def delete(object)
-    @hash.delete(object)
-    self
+  def delete(object) : T?
+    return nil if !object.is_a? T
+    @hash.delete(object) { return nil }
+    object
```

Comment:
Returning the argument as is indeed has no benefit to just returning `Bool`.

The potential advantage of returning `T?` is to return the *original object* that was in the set. This can be useful if you have objects with partial equality, that is you can reconstruct it based on data you have and that makes it equal (say based on opaque identifier  given to API consumers), while you need additional context data contained in the original object but not available in the current context (say a database PK).

Reply:
Yes that's right, I will change that

Reply:
That's not possible with the current `Hash` API, changing the method to return a `Bool`.